openstudiocore/src/model/MeterCustom.cpp

```diff
@@ -0,0 +1,298 @@
+  }
+
+  MeterCustom_Impl::MeterCustom_Impl(const openstudio::detail::WorkspaceObject_Impl& other,
+                                     Model_Impl* model,
+                                     bool keepHandle)
+    : ModelObject_Impl(other, model, keepHandle)
+  {
+    OS_ASSERT(other.iddObject().type() == MeterCustom::iddObjectType());
+  }
+
+  MeterCustom_Impl::MeterCustom_Impl(const MeterCustom_Impl& other,
+                                     Model_Impl* model,
+                                     bool keepHandle)
+    : ModelObject_Impl(other,model,keepHandle)
+  {}
+
+  // This logically doesn't produce any output variables
+  const std::vector<std::string>& MeterCustom_Impl::outputVariableNames() const
+  {
+    static std::vector<std::string> result;
+    if (result.empty()){
+    }
+    LOG(Info, "Meter:Custom does not produce any output variables.");
+    return result;
+  }
+
+  IddObjectType MeterCustom_Impl::iddObjectType() const {
+    return MeterCustom::iddObjectType();
+  }
+
+  // fuel Type
+  boost::optional<std::string> MeterCustom_Impl::fuelType() const {
+    return getString(OS_Meter_CustomFields::FuelType,true);
+  }
+
+  bool MeterCustom_Impl::setFuelType(boost::optional<std::string> fuelType) {
+    bool result(false);
+    if (fuelType) {
+      result = setString(OS_Meter_CustomFields::FuelType, fuelType.get());
+    }
+    else {
+      resetFuelType();
+      result = true;
+    }
+    return result;
+  }
+
+  void MeterCustom_Impl::resetFuelType() {
+    bool result = setString(OS_Meter_CustomFields::FuelType, "");
+    OS_ASSERT(result);
+  }
+
+
+  // Add a new (Key, Var) group
+  bool MeterCustom_Impl::addKeyVarGroup(const std::string& keyName, const std::string& outputVariableorMeterName) {
+    WorkspaceExtensibleGroup eg = getObject<ModelObject>().pushExtensibleGroup().cast<WorkspaceExtensibleGroup>();
+
+    bool temp = eg.setString(OS_Meter_CustomExtensibleFields::KeyName, keyName);
+    bool ok = eg.setString(OS_Meter_CustomExtensibleFields::OutputVariableorMeterName, outputVariableorMeterName);
+
+    if (temp) {
+      temp = ok;
+    }
+
+    if (!temp) {
+      getObject<ModelObject>().eraseExtensibleGroup(eg.groupIndex());
+      return temp;
+    }
+    return temp;
+  }
+
+  // Remove the (Key, Var) group at given index
+  bool MeterCustom_Impl::removeKeyVarGroup(unsigned groupIndex) {
+    bool ok = false;
+    unsigned numberofDataPairs = numExtensibleGroups();
+    // Only delete if the index is smaller than the number of groups
+    if (groupIndex < numberofDataPairs) {
+      getObject<ModelObject>().eraseExtensibleGroup(groupIndex);
+      ok = true;
+    }
+
+    return ok;
+  }
+
+  // Remove all the (Key, Var) groups
+  void MeterCustom_Impl::removeAllKeyVarGroups() {
+    getObject<ModelObject>().clearExtensibleGroups();
+  }
+
+  // Return a vector of (Key, Var) pairs
+  std::vector< std::pair<std::string, std::string> > MeterCustom_Impl::keyVarGroups() {
```

Comment:
@jmarrec I would prefer to define a simple class:

`struct MeterCustomKeyVarGroup{
  std::string key;
  std::string variable;
}`

than use `std::pair<string, string>` it will be much easier to work with in SWIG